src/pending.js

```diff
@@ -0,0 +1,144 @@
+var _ = require('lodash');
+var debug = require('debug')('stats-collector:pending');
+var taskcluster = require('taskcluster-client');
+
+const FLUSH_INTERVAL = 3; // seconds
+
+// check for pending tasks that aren't really pending with this frequency; should
+// be smaller than MIN_CHECK_AGE
+const CHECK_INTERVAL = 1; // seconds
+
+// only check tasks pending at least this long; this should be lower than any alert
+// thresholds for pending tasks.
+const MIN_CHECK_AGE = 300; // seconds
+
+class PendingCollector {
+  // Monitor the longest-pending task for each workerType
+
+  constructor ({monitor, listener}) {
+    // mappings from task key to pending time, keyed by workerType; an empty list
+    // here is significant in that it means there is nothing pending for that
+    // workerType
+    this.pendingTasks = {};
+
+    // workerTypes for which we have seen a single task both enter and exit the
+    // pending state, and thus have a reasonable expectation of knowing the
+    // total pending time for, even if the stats-collector process has just
+    // started.
+    this.readyWorkerTypes = {};
+
+    this.queue = new taskcluster.Queue();
+    this.listener = listener;
+    this.monitor = monitor;
+  };
+
+  start () {
+    this.listener.on('task-message', ({action, payload}) => {
+      try {
+        let taskKey = `${payload.status.taskId}/${payload.runId}`;
+        let workerType = payload.status.workerType;
+        let isPending = action === 'task-pending';
+        let scheduled = payload.status.runs[payload.runId].scheduled;
+
+        this.update(workerType, taskKey, isPending, scheduled);
+      } catch (err) {
+        debug('Failed to process message %s with error: %s, as JSON: %j',
+              action, err, err, err.stack);
+      }
+    });
+
+    // set up periodic flushes
+    this.periodically(FLUSH_INTERVAL * 1000, 'flush', () => {
+      return this.flush(new Date().getTime());
+    });
+
+    // set up periodic checks of the oldest task
+    this.periodically(CHECK_INTERVAL * 1000, 'check', () => {
+      return this.check(new Date().getTime());
+    });
```

Comment:
hmm... I would run check until the earliest task is pending... right before flush() every time.. And then only flush every 60s...

But I guess this works too... there is just higher risk of reporting something wrong. Though it probably won't be too wrong.